### PR TITLE
Feed source table refactor

### LINKED ISSUE #422: Feed source table view
Replace the current list view with a more table-like UI to do the following items. We should draw inspiration/steal code from the alerts list.
- [x] Show service start and service end dates for current service for all agencies;
- [ ] Show if data for future service is available in TDM and when it was loaded for all agencies - **more info needed** (should this table show publishedValidation or latestValidation). If MTC extension is enabled, use publishedValidation and include info about "future service", if extension is not enabled, default to latesetValidation.
- [x] highlight service end date for agencies depending on expiration
-- highlight service end date for agencies with data expiring in > 20 days  in color Green;
-- highlight service end date for agencies with data expiring in < 20 days  in color Yellow;
-- highlight service end date for  agencies with data expiring in < 5 days  in color Red;
-- highlight service end date for  agencies with data expired in Black;
- [x] Automatically send an email to users subscribed to a feed two weeks before the latest feed expires
- [x] Remove the existing Calendar icon indicating service status
- [x] Move other icons (private, edit, error status, feed link) in separate columns (instead of current placement around the agency names)
- [x] Sort feeds by start, end date, or number of errors

### PR DESCRIPTION
**Pull request by evansiroky**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JSDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

Refactor the Project view's Feedsource table and toolbar. Seeks to resolve #422. Also add pinned deployments to projects. This PR depends on https://github.com/conveyal/datatools-server/pull/191.

This will work with any configuration, but works especially well if deployments are enable and if the application is able to publish feed versions.

#### Feed Source Table Refactor

The biggest part of this PR involves refactoring the Feed Source Table in the project viewer to show more information. It is now able to display basic stats about the current feed version and also allows for the comparison to another feed version that is either in a deployment or has been published.

Internally, the ProjectViewer was refactored to now render a new connected component FeedSourceTable that now renders the Panel, Toolbar and Table. The ProjectFeedListToolbar is now also connected to avoid passing a zillion props down the line. The toolbar has been refactored to allow sorting of the table, showing counts by version status, and moving some action buttons into a dropdown. The old FeedSourceTable component now renders an html table instead of bootstrap ListItems. A table header is now rendered and each row is rendered as the connected component FeedSourceTableRow.

A fair amount of new logic in the actions, reducers and util function was also added in order to properly render the FeedSourceTable. There are a few new fields in the projects reducer and supporting actions. Also, the onProjectViewerMount action had more logic added to it so that additional data that is needed to make comparisons is loaded right away.

#### Pinned Deployments

It is now possible to "pin" a deployment. When a deployment is pinned it is highlighted and shown as the first deployment in the projects list. Futhermore, if a project has a pinned deployment, the FeedSourceTable will compare the versions in that pinned deployment to the latest feedsource versions.

#### Testing!

This PR introduces a lot of new mock data and two types of testing methods that I hope are simple to make, easy to understand and covering a lot of code. The mocks are now organized within a directory in the `__tests__/test-utils` folder.

The first type of test is of the actions and reducers all at once. The tests in `lib/manager/actions/__tests__/projects.js` make tests that check if the end result of an action updates the store as expected. The second type of test is of some container components. Each test in `lib/manager/containers/__tests__` will make sure that a connected component will render properly given an underlying store state and props provided to the connected component. Testing containers instead of the underlying components also has the added benefit of often times invoking selectors and util functions. Each of these types of tests hopefully are well received for getting the most bang for our buck to help us make tests with limited time. The downside of these approaches is that they heavily employ snapshots which at times can generate massive amounts of data that can provide false positives. I did briefly inspect each snapshot and feel that they contain the proper data to the best of my knowledge.

### CODE REVIEW

**Comment by evansiroky** [comment]
Just created this PR to have a place for feedback on this. FYI, there is another branch at https://github.com/conveyal/datatools-ui/tree/compact-feedsourcetable that builds on this PR and allows a compact view of the table.

**Comment by evansiroky** [comment]
Just assigned this back to myself as we've had internal discussion on how to redesign the table.

**Comment by evansiroky** [comment]
Description updated. This should be ready for review.

**Comment by landonreed** [comment]
@evansiroky to make some refinements to snapshots and sidebar props.

File: __tests__/test-utils/mock-data/manager.js [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+    otpCommit: null,
+    project,
+    projectBounds: (null: ?Bounds), // not sure why this flow type annotation is needed (see https://git.io/fjJCm)
+    routerId: (null: ?string), // not sure why this flow type annotation is needed (see https://git.io/fjJCm)
+    user: (null: ?any) // not sure why this flow type annotation is needed (see https://git.io/fjJCm)
+  }
+}
+
+// a mock project without any feeds or deployments
+export const mockProject = {
```
Comment (landonreed) [inline __tests__/test-utils/mock-data/manager.js]: It seems like this should have a type defined. Is there a reason why it doesn't?

File: __tests__/test-utils/mock-data/manager.js [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    stairsReluctance: null,
+    updaters: null,
+    walkSpeed: null
+  },
+  useCustomOsmBounds: false,
+  user: null
+}
+
+// a mock feed with a version
+export const mockFeedWithVersion = {
```
Comment (landonreed) [inline __tests__/test-utils/mock-data/manager.js]: It seems like this should have a type defined. Is there a reason why it doesn't?

File: __tests__/test-utils/mock-data/manager.js [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+  publishedVersionId: null,
+  retrievalMethod: 'MANUALLY_UPLOADED',
+  s3Url: null,
+  snapshotVersion: null,
+  url: 'http://mdtrip.org/googletransit/AnnapolisTransit/google_transit.zip',
+  user: null
+}
+
+// a mock feed with no versions
+export const mockFeedWithoutVersion = {
```
Comment (landonreed) [inline __tests__/test-utils/mock-data/manager.js]: It seems like this should have a type defined. Is there a reason why it doesn't?

File: __tests__/test-utils/mock-data/manager.js [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+  publishedVersionId: null,
+  retrievalMethod: 'FETCHED_AUTOMATICALLY',
+  s3Url: null,
+  snapshotVersion: null,
+  url: null,
+  user: null
+}
+
+// a mock feedversion that has validation data
+export const mockFeedVersion = {
```
Comment (landonreed) [inline __tests__/test-utils/mock-data/manager.js]: It seems like this should have a type defined.

File: __tests__/test-utils/mock-data/manager.js (hunk 1 above)
Comment (evansiroky) [reply __tests__/test-utils/mock-data/manager.js]: The typing of these mock objects are checked when they are sent to various methods as function arguments. If the object structure doesn't match the argument type, a flow error occurs.

File: __tests__/test-utils/mock-data/manager.js (hunk 2 above)
Comment (evansiroky) [reply __tests__/test-utils/mock-data/manager.js]: The typing of these mock objects are checked when they are sent to various methods as function arguments. If the object structure doesn't match the argument type, a flow error occurs.

File: __tests__/test-utils/mock-data/manager.js (hunk 3 above)
Comment (evansiroky) [reply __tests__/test-utils/mock-data/manager.js]: The typing of these mock objects are checked when they are sent to various methods as function arguments. If the object structure doesn't match the argument type, a flow error occurs.

File: __tests__/test-utils/mock-data/manager.js (hunk 4 above)
Comment (evansiroky) [reply __tests__/test-utils/mock-data/manager.js]: The typing of these mock objects are checked when they are sent to various methods as function arguments. If the object structure doesn't match the argument type, a flow error occurs.

**Comment by evansiroky** [comment]
I fixed the sidebar component so it doesn't result in massive snapshots. Also, I made this branch depend on the usage of a datatools-server that has the changes in https://github.com/conveyal/datatools-server/pull/195.

**Comment by evansiroky** [comment]
Assigning back to me. I'll be working on adding the ability to download a csv of the table.

**Comment by evansiroky** [comment]
Just added ability to download CSV of the FeedSourceTable.

**Pull request merged**